File: packages/adapters/src/venues/kucoin.ts
import {
  type AssetClass,
  canonicalBase,
  classifyNonCrypto,
  type FundingEvent,
  type FundingSnapshot,
  inferIntervalHours,
  type LeverageTier,
} from "@ai-rates/core";
import { CircuitOpenError } from "../http";
import { marketRef, mul, num } from "../parse";
import type { SnapshotBatch, VenueAdapter } from "../types";

const VENUE_ID = "kucoin";
const BASE_URL = "https://api-futures.kucoin.com";
const MS_PER_HOUR = 3_600_000;
const HISTORY_PAGE = 100;
const MAX_PAGES = 50;
const RISK_LIMIT_RETRIES = 1;
const RISK_LIMIT_BACKOFF_MS = 300;
const OK = "200000";

/** KuCoin's type code for perpetual swaps; "FFICSX" is dated futures. */
const PERPETUAL = "FFWCSX";

export interface KucoinEnvelope<T> {
  code: string;
  msg?: string;
  data: T;
}

export interface KucoinContract {
  symbol: string;
  type: string;
  status: string;
  isInverse: boolean;
  /** Rate for the current funding period. */
  fundingFeeRate: number | null;
  /** ms; null on some symbols, where `fundingRateGranularity` still holds the interval. */
  currentFundingRateGranularity: number | null;
  fundingRateGranularity: number | null;
  nextFundingRateDateTime: number | null;
  /** Rate settled at the previous funding time. */
  lastTimeFundingRate: number | null;
  markPrice: number | null;
  indexPrice: number | null;
  /** Lots, as a string. */
  openInterest: string;
  /** Base units per lot for linear contracts. */
  multiplier: number;
  /** Headline leverage; /contracts/risk-limit/{symbol} is the precise, size-aware source. */
  maxLeverage?: number | null;
  turnoverOf24h: number | null;
  /** What the contract tracks: "CRYPTO", "STOCK", "METAL" or "COMMODITY". */
  assetClass?: string;
}

/**
 * The class KuCoin declares for a contract, from `assetClass` on /api/v1/contracts/active.
 *
 * Live 2026-09-14, 687 contracts: CRYPTO 532, STOCK 146, METAL 6, COMMODITY 3. It separates
 * BBXUSDTM and QNTXUSDTM (stocks) from BBUSDTM and QNTUSDTM (crypto), and files CL, BZ and NATGAS
 * as commodities. `marketType` is NOT a substitute: it reads CRYPTO on every METAL and COMMODITY
 * row. PAXG and XAUT are declared METAL and returned to crypto by `marketRef`. A missing field is
 * undeclared, so crypto; a value this does not know still says not-crypto, so the base tables
 * settle which class.
 */
export function kucoinAssetClass(assetClass: string | undefined, base: string): AssetClass {
  switch (assetClass ?? "CRYPTO") {
    case "CRYPTO":
      return "crypto";
    case "STOCK":
      return "equity";
    case "METAL":
    case "COMMODITY":
      return "commodity";
    default:
      return classifyNonCrypto(canonicalBase(base));
  }
}

export interface KucoinRiskLimit {
  symbol: string;
  level: number;
  /** Quote notional, inclusive upper bound of the band. */
  maxRiskLimit: number;
  /** Quote notional; equals the previous level's `maxRiskLimit`, so bands are already contiguous. */
  minRiskLimit: number;
  maxLeverage: number;
  initialMargin: number;
  maintainMargin: number;
}

export interface KucoinFundingHistoryItem {
  symbol: string;
  fundingRate: number;
  timepoint: number;
}

function unwrap<T>(json: KucoinEnvelope<T>, what: string): T {
  if (json.code !== OK) throw new Error(`kucoin ${what}: ${json.code} ${json.msg ?? ""}`);
  return json.data;
}

/**
 * KuCoin answers `data: null` rather than `[]` when a window holds no rows, which the backfill hits
 * constantly: a symbol listed last week has nothing 90 days back. That is an empty result, not a
 * failure, and returning [] lets the caller mark the market exhausted instead of retrying forever.
 */
function unwrapList<T>(json: KucoinEnvelope<T[] | null>, what: string): T[] {
  return unwrap(json, what) ?? [];
}

function granularityMs(contract: KucoinContract): number | null {
  const ms = contract.currentFundingRateGranularity ?? contract.fundingRateGranularity;
  return ms !== null && ms > 0 ? ms : null;
}

export function parseKucoinSnapshots(
  json: KucoinEnvelope<KucoinContract[] | null>,
  now: number,
): SnapshotBatch {
  const snapshots: FundingSnapshot[] = [];
  const settled: FundingEvent[] = [];

  for (const contract of unwrapList(json, "contracts")) {
    if (contract.type !== PERPETUAL || contract.isInverse || contract.status !== "Open") continue;
    const intervalMs = granularityMs(contract);
    const rate = num(contract.fundingFeeRate);
    if (intervalMs === null || rate === null) continue;

    const hours = intervalMs / MS_PER_HOUR;
    const { base } = marketRef(VENUE_ID, contract.symbol);
    const ref = marketRef(VENUE_ID, contract.symbol, {
      assetClass: kucoinAssetClass(contract.assetClass, base),
    });
    const nextFundingAt = num(contract.nextFundingRateDateTime);
    const markPrice = num(contract.markPrice);
    snapshots.push({
      ...ref,
      observedAt: now,
      rate,
      basisHours: hours,
      intervalHours: hours,
      nextFundingAt,
      kind: "predicted",
      markPrice,
      indexPrice: num(contract.indexPrice),
      openInterestUsd: mul(num(contract.openInterest), num(contract.multiplier), markPrice),
      volume24hUsd: num(contract.turnoverOf24h),
      maxLeverage: num(contract.maxLeverage),
    });

    const lastRate = num(contract.lastTimeFundingRate);
    if (lastRate !== null && nextFundingAt !== null) {
      settled.push({
        ...ref,
        settledAt: nextFundingAt - intervalMs,
        rate: lastRate,
        basisHours: hours,
        markPrice: null,
      });
    }
  }
  return { snapshots, settled };
}

/**
 * KuCoin's risk ladders, one per symbol.
 *
 * It publishes both bounds, and level n's `minRiskLimit` equals level n-1's `maxRiskLimit`, so no
 * floor has to be inferred the way Bybit's and Gate's do. Bounds are quote notional, not lots:
 * `initialMargin` 0.008 is exactly 1/125 matching `maxLeverage`, and read as lots XBTUSDTM's first
 * band would be a $19m position at 125x, which no venue offers.
 */
export function parseKucoinRiskLimits(rows: readonly KucoinRiskLimit[]): LeverageTier[] {
  const bySymbol = new Map<string, KucoinRiskLimit[]>();
  for (const row of rows) {
    const existing = bySymbol.get(row.symbol);
    if (existing) existing.push(row);
    else bySymbol.set(row.symbol, [row]);
  }

  const ladders: LeverageTier[] = [];
  for (const [symbol, symbolRows] of bySymbol) {
    const ladder: LeverageTier[] = [];
    let usable = true;

    for (const row of [...symbolRows].sort((a, b) => a.level - b.level)) {
      const lower = num(row.minRiskLimit);
      const upper = num(row.maxRiskLimit);
      const imr = num(row.initialMargin);
      const maxLeverage = num(row.maxLeverage);
      if (
        lower === null ||
        upper === null ||
        upper <= lower ||
        imr === null ||
        imr <= 0 ||
        maxLeverage === null ||
        maxLeverage <= 0
      ) {
        usable = false;
        break;
      }
      ladder.push({
        venueId: VENUE_ID,
        venueSymbol: symbol,
        tier: row.level,
        lowerNotionalUsd: lower,
        upperNotionalUsd: upper,
        imr,
        mmr: num(row.maintainMargin),
        maxLeverage,
      });
    }
    if (usable) ladders.push(...ladder);
  }
  return ladders;
}

/** Settled funding events for one symbol, oldest first. */
export function parseKucoinFundingHistory(
  json: KucoinEnvelope<KucoinFundingHistoryItem[] | null>,
  fallbackHours: number,
): FundingEvent[] {
  const points = unwrapList(json, "funding history")
    .map((item) => ({
      symbol: item.symbol,
      settledAt: num(item.timepoint),
      rate: num(item.fundingRate),
    }))
    .filter(
      (p): p is { symbol: string; settledAt: number; rate: number } =>
        p.settledAt !== null && p.rate !== null,
    )
    .sort((a, b) => a.settledAt - b.settledAt);

  const basisHours = inferIntervalHours(points.map((p) => p.settledAt)) ?? fallbackHours;
  return points.map((p) => ({
    ...marketRef(VENUE_ID, p.symbol),
    settledAt: p.settledAt,
    rate: p.rate,
    basisHours,
    markPrice: null,
  }));
}

export const kucoinAdapter: VenueAdapter = {
  venueId: VENUE_ID,
  minIntervalMs: 150,

  async fetchSnapshots(client, now) {
    const json = await client.getJson<KucoinEnvelope<KucoinContract[]>>(
      `${BASE_URL}/api/v1/contracts/active`,
    );
    return parseKucoinSnapshots(json, now);
  },

  /**
   * The only venue here with no bulk form — asking without a symbol answers 404000 — so this is
   * one call per market: ~680 requests at 150ms spacing, about 100 seconds once a day. That holds
   * KuCoin's shared client long enough to delay a snapshot cycle or two, which is the price of
   * having ladders at all for this venue.
   */
  async fetchLeverageTiers(client) {
    const contracts = unwrapList(
      await client.getJson<KucoinEnvelope<KucoinContract[] | null>>(
        `${BASE_URL}/api/v1/contracts/active`,
      ),
      "contracts",
    );
    const symbols = contracts
      .filter((c) => c.type === PERPETUAL && !c.isInverse && c.status === "Open")
      .map((c) => c.symbol);

    const rows: KucoinRiskLimit[] = [];
    let complete = true;

    for (const symbol of symbols) {
      let fetched: KucoinRiskLimit[] | null = null;
      for (let attempt = 0; fetched === null && attempt <= RISK_LIMIT_RETRIES; attempt++) {
        try {
          fetched = unwrapList(
            await client.getJson<KucoinEnvelope<KucoinRiskLimit[] | null>>(
              `${BASE_URL}/api/v1/contracts/risk-limit/${encodeURIComponent(symbol)}`,
            ),
            "risk limit",
          );
        } catch (error) {
          if (error instanceof CircuitOpenError) {
            return { tiers: parseKucoinRiskLimits(rows), complete: false };
          }
          if (attempt < RISK_LIMIT_RETRIES) {
            await new Promise((resolve) => setTimeout(resolve, RISK_LIMIT_BACKOFF_MS << attempt));
          }
        }
      }
      if (fetched === null) complete = false;
      else rows.push(...fetched);
    }
    return { tiers: parseKucoinRiskLimits(rows), complete };
  },

  async fetchFundingHistory(client, venueSymbol, fromMs, toMs) {
    const items: KucoinFundingHistoryItem[] = [];
    let to = toMs;
    for (let page = 0; page < MAX_PAGES && to >= fromMs; page++) {
      const data = unwrapList(
        await client.getJson<KucoinEnvelope<KucoinFundingHistoryItem[] | null>>(
          `${BASE_URL}/api/v1/contract/funding-rates?symbol=${encodeURIComponent(venueSymbol)}&from=${fromMs}&to=${to}`,
        ),
        "funding history",
      );
      items.push(...data);
      if (data.length < HISTORY_PAGE) break;
      to = Math.min(...data.map((i) => i.timepoint)) - 1;
    }

    let fallbackHours = 8;
    if (items.length < 2) {
      const contract = unwrap(
        await client.getJson<KucoinEnvelope<KucoinContract>>(
          `${BASE_URL}/api/v1/contracts/${encodeURIComponent(venueSymbol)}`,
        ),
        "contract",
      );
      const ms = granularityMs(contract);
      if (ms !== null) fallbackHours = ms / MS_PER_HOUR;
    }
    const unique = [...new Map(items.map((i) => [i.timepoint, i])).values()];
    return parseKucoinFundingHistory({ code: OK, data: unique }, fallbackHours);
  },
};

File: apps/collector-go/internal/adapters/kucoin/kucoin.go
// Package kucoin parses KuCoin Futures' perpetual market data.
//
// Ported from packages/adapters/src/venues/kucoin.ts. Two things about this venue differ from every
// other adapter here and are easy to get wrong:
//
//   - Symbols carry a trailing M (XBTUSDTM), and XBT is KuCoin's spelling of BTC. Both are handled
//     by core.ParseVenueSymbol and core.CanonicalBase, so the symbol is never split here.
//   - The funding interval is reported in MILLISECONDS (28,800,000 for 8h), not minutes as Bybit
//     does or hours as Bitget does.
package kucoin

import (
	"fmt"
	"sort"

	"github.com/belyjelli/ai-rates/collector/internal/adapters"
	"github.com/belyjelli/ai-rates/collector/internal/core"
)

const VenueID = "kucoin"

const (
	msPerHour = 3_600_000.0

	// ok is KuCoin's success code. Anything else carries a message and no usable data, including
	// rate limits, which arrive as 429000 inside an HTTP 200.
	ok = "200000"

	// perpetual is KuCoin's type code for perpetual swaps; "FFICSX" is dated futures.
	perpetual = "FFWCSX"
)

// Envelope is KuCoin's uniform response wrapper.
type Envelope[T any] struct {
	Code string `json:"code"`
	Msg  string `json:"msg"`
	Data T      `json:"data"`
}

// unwrap returns the payload, or an error when KuCoin reported a failure code.
//
// A list payload needs no separate accessor the way the TypeScript's `unwrapList` does: KuCoin
// answers `data: null` rather than `[]` when a window holds no rows, which the backfill hits
// constantly — a symbol listed last week has nothing 90 days back — and JSON null decodes to a nil
// slice, which ranges as empty. That is an empty result, not a failure, and returning it lets the
// caller mark the market exhausted instead of retrying forever. On the TypeScript side the same null
// threw "Spread syntax requires ...iterable" in production until it was handled.
func (e Envelope[T]) unwrap(what string) (T, error) {
	if e.Code != ok {
		var zero T
		return zero, fmt.Errorf("kucoin %s: %s %s", what, e.Code, e.Msg)
	}
	return e.Data, nil
}

type Contract struct {
	Symbol string `json:"symbol"`
	Type   string `json:"type"`
	Status string `json:"status"`
	// IsInverse marks the coin-margined contracts, which this adapter skips.
	IsInverse bool `json:"isInverse"`
	// FundingFeeRate is the rate for the current funding period.
	FundingFeeRate adapters.Num `json:"fundingFeeRate"`
	// CurrentFundingRateGranularity is the interval in MILLISECONDS; absent on some symbols, where
	// FundingRateGranularity still holds it.
	CurrentFundingRateGranularity adapters.Num `json:"currentFundingRateGranularity"`
	FundingRateGranularity        adapters.Num `json:"fundingRateGranularity"`
	NextFundingRateDateTime       adapters.Num `json:"nextFundingRateDateTime"`
	// LastTimeFundingRate is the rate settled at the previous funding time.
	LastTimeFundingRate adapters.Num `json:"lastTimeFundingRate"`
	MarkPrice           adapters.Num `json:"markPrice"`
	IndexPrice          adapters.Num `json:"indexPrice"`
	// OpenInterest is LOTS, quoted as a string. Multiplier converts a lot to base units.
	OpenInterest adapters.Num `json:"openInterest"`
	// Multiplier is base units per lot for linear contracts.
	Multiplier adapters.Num `json:"multiplier"`
	// MaxLeverage is the headline figure; /contracts/risk-limit/{symbol} is the precise, size-aware
	// source.
	MaxLeverage   adapters.Num `json:"maxLeverage"`
	TurnoverOf24h adapters.Num `json:"turnoverOf24h"`
	// AssetClass is what the contract tracks: "CRYPTO", "STOCK", "METAL" or "COMMODITY".
	AssetClass string `json:"assetClass"`
}

// AssetClassFor is the class KuCoin declares for a contract, from `assetClass` on
// /api/v1/contracts/active.
//
// Live 2026-09-14, 687 contracts: CRYPTO 532, STOCK 146, METAL 6, COMMODITY 3. It separates
// BBXUSDTM and QNTXUSDTM (stocks) from BBUSDTM and QNTUSDTM (crypto), and files CL, BZ and NATGAS
// as commodities. `marketType` is NOT a substitute: it reads CRYPTO on every METAL and COMMODITY
// row. PAXG and XAUT are declared METAL and returned to crypto by MarketRefFor. A missing field is
// undeclared, so crypto; a value this does not know still says not-crypto, so the base tables
// settle which class.
//
// Go cannot tell an absent string from an empty one, so "" is read as undeclared. The TypeScript's
// `?? "CRYPTO"` would send a literal empty string down the unknown branch instead; KuCoin publishes
// no such value, and reading a field it does not send as crypto is the same default every other
// undeclared venue gets.
func AssetClassFor(declared, base string) core.AssetClass {
	switch declared {
	case "", "CRYPTO":
		return core.ClassCrypto
	case "STOCK":
		return core.ClassEquity
	case "METAL", "COMMODITY":
		return core.ClassCommodity
	default:
		return core.ClassifyNonCrypto(core.CanonicalBase(base))
	}
}

type RiskLimit struct {
	Symbol string       `json:"symbol"`
	Level  adapters.Num `json:"level"`
	// MaxRiskLimit is quote notional, the inclusive upper bound of the band.
	MaxRiskLimit adapters.Num `json:"maxRiskLimit"`
	// MinRiskLimit is quote notional; it equals the previous level's MaxRiskLimit, so bands are
	// already contiguous.
	MinRiskLimit   adapters.Num `json:"minRiskLimit"`
	MaxLeverage    adapters.Num `json:"maxLeverage"`
	InitialMargin  adapters.Num `json:"initialMargin"`
	MaintainMargin adapters.Num `json:"maintainMargin"`
}

type FundingHistoryItem struct {
	Symbol      string       `json:"symbol"`
	FundingRate adapters.Num `json:"fundingRate"`
	Timepoint   adapters.Num `json:"timepoint"`
}

// granularityMs is the funding interval in MILLISECONDS, or nil when the contract publishes none.
func granularityMs(contract Contract) *float64 {
	ms := contract.CurrentFundingRateGranularity
	if !ms.OK {
		ms = contract.FundingRateGranularity
	}
	if !ms.OK || ms.Val <= 0 {
		return nil
	}
	v := ms.Val
	return &v
}

// ParseSnapshots normalises one /contracts/active response into snapshots and the settlements that
// arrive free with it.
func ParseSnapshots(env Envelope[[]Contract], now int64) (core.SnapshotBatch, error) {
	contracts, err := env.unwrap("contracts")
	if err != nil {
		return core.SnapshotBatch{}, err
	}

	snapshots := make([]core.FundingSnapshot, 0, len(contracts))
	settled := make([]core.FundingEvent, 0, len(contracts))

	for _, contract := range contracts {
		if contract.Type != perpetual || contract.IsInverse || contract.Status != "Open" {
			continue
		}
		intervalMs := granularityMs(contract)
		if intervalMs == nil || !contract.FundingFeeRate.OK {
			continue
		}

		hours := *intervalMs / msPerHour
		base := adapters.MarketRefFor(VenueID, contract.Symbol, adapters.Overrides{}).Base
		class := AssetClassFor(contract.AssetClass, base)
		ref := adapters.MarketRefFor(VenueID, contract.Symbol, adapters.Overrides{AssetClass: &class})

		// PositiveMs rather than a plain number, so a venue reporting 0 for "no next funding" is read
		// as absent. Carrying a zero through would date the settlement below at minus one interval.
		nextFundingAt := contract.NextFundingRateDateTime.PositiveMs()
		markPrice := contract.MarkPrice.Ptr()

		interval := hours
		snapshots = append(snapshots, core.FundingSnapshot{
			MarketRef:     ref,
			ObservedAt:    now,
			Rate:          contract.FundingFeeRate.Val,
			BasisHours:    hours,
			IntervalHours: &interval,
			NextFundingAt: nextFundingAt,
			Kind:          core.KindPredicted,
			MarkPrice:     markPrice,
			IndexPrice:    contract.IndexPrice.Ptr(),
			// Open interest is LOTS: 11,784,994 lots of XBTUSDTM at 0.001 BTC a lot is 11,785 BTC.
			// Reading the lot count as coin would be three orders of magnitude out.
			OpenInterestUSD: adapters.Mul(contract.OpenInterest.Ptr(), contract.Multiplier.Ptr(), markPrice),
			Volume24hUSD:    contract.TurnoverOf24h.Ptr(),
			MaxLeverage:     contract.MaxLeverage.Ptr(),
		})

		if contract.LastTimeFundingRate.OK && nextFundingAt != nil {
			settled = append(settled, core.FundingEvent{
				MarketRef:  ref,
				SettledAt:  *nextFundingAt - int64(*intervalMs),
				Rate:       contract.LastTimeFundingRate.Val,
				BasisHours: hours,
				MarkPrice:  nil,
			})
		}
	}
	return core.SnapshotBatch{Snapshots: snapshots, Settled: settled}, nil
}

// ParseRiskLimits converts KuCoin's risk ladders, one per symbol.
//
// KuCoin publishes BOTH bounds, and level n's MinRiskLimit equals level n-1's MaxRiskLimit, so no
// floor has to be inferred the way Bybit's and Gate's do. Bounds are quote notional, not lots:
// InitialMargin 0.008 is exactly 1/125, matching MaxLeverage, and read as lots XBTUSDTM's first band
// would be a $19m position at 125x, which no venue offers.
//
// Symbols keep the order they first appear in, which a Go map would not: the TypeScript walks a Map
// in insertion order, and a ladder's position in the output is the venue's own ordering.
func ParseRiskLimits(rows []RiskLimit) []core.LeverageTier {
	order := make([]string, 0, len(rows))
	bySymbol := make(map[string][]RiskLimit, len(rows))
	for _, row := range rows {
		if _, seen := bySymbol[row.Symbol]; !seen {
			order = append(order, row.Symbol)
		}
		bySymbol[row.Symbol] = append(bySymbol[row.Symbol], row)
	}

	ladders := make([]core.LeverageTier, 0, len(rows))
	for _, symbol := range order {
		symbolRows := make([]RiskLimit, len(bySymbol[symbol]))
		copy(symbolRows, bySymbol[symbol])
		sort.SliceStable(symbolRows, func(i, j int) bool { return symbolRows[i].Level.Val < symbolRows[j].Level.Val })

		ladder := make([]core.LeverageTier, 0, len(symbolRows))
		usable := true
		for _, row := range symbolRows {
			if !row.MinRiskLimit.OK || !row.MaxRiskLimit.OK || row.MaxRiskLimit.Val <= row.MinRiskLimit.Val ||
				!row.InitialMargin.OK || row.InitialMargin.Val <= 0 ||
				!row.MaxLeverage.OK || row.MaxLeverage.Val <= 0 {
				// An unreadable tier drops the WHOLE ladder, as on Bybit and OKX: keeping the rest
				// would stretch a neighbouring band across the gap and quote confident margin for a
				// range nothing verified.
				usable = false
				break
			}
			upper := row.MaxRiskLimit.Val
			ladder = append(ladder, core.LeverageTier{
				VenueID:          VenueID,
				VenueSymbol:      symbol,
				Tier:             int(row.Level.Val),
				LowerNotionalUSD: row.MinRiskLimit.Val,
				UpperNotionalUSD: &upper,
				IMR:              row.InitialMargin.Val,
				MMR:              row.MaintainMargin.Ptr(),
				MaxLeverage:      row.MaxLeverage.Val,
			})
		}
		if usable {
			ladders = append(ladders, ladder...)
		}
	}
	return ladders
}

// ParseFundingHistory returns settled funding events for one symbol, oldest first.
func ParseFundingHistory(env Envelope[[]FundingHistoryItem], fallbackHours float64) ([]core.FundingEvent, error) {
	rows, err := env.unwrap("funding history")
	if err != nil {
		return nil, err
	}

	type point struct {
		symbol    string
		settledAt int64
		rate      float64
	}
	points := make([]point, 0, len(rows))
	for _, item := range rows {
		at := item.Timepoint.PositiveMs()
		if at == nil || !item.FundingRate.OK {
			continue
		}
		points = append(points, point{item.Symbol, *at, item.FundingRate.Val})
	}
	sort.SliceStable(points, func(i, j int) bool { return points[i].settledAt < points[j].settledAt })

	stamps := make([]int64, len(points))
	for i, p := range points {
		stamps[i] = p.settledAt
	}
	basisHours := fallbackHours
	if inferred := core.InferIntervalHours(stamps); inferred != nil {
		basisHours = *inferred
	}

	events := make([]core.FundingEvent, 0, len(points))
	for _, p := range points {
		events = append(events, core.FundingEvent{
			MarketRef:  adapters.MarketRefFor(VenueID, p.symbol, adapters.Overrides{}),
			SettledAt:  p.settledAt,
			Rate:       p.rate,
			BasisHours: basisHours,
			MarkPrice:  nil,
		})
	}
	return events, nil
}
